service/storage_proxy.cc

```diff
@@ -1381,19 +1381,19 @@ class abstract_write_response_handler : public seastar::enable_shared_from_this<
     std::chrono::microseconds calculate_delay(db::view::update_backlog backlog) {
         constexpr auto delay_limit_us = 1000000;
         auto adjust = [] (float x) { return x * x * x; };
-        auto budget = std::max(storage_proxy::clock_type::duration(0),
-            _expire_timer.get_timeout() - storage_proxy::clock_type::now());
         std::chrono::microseconds ret(uint32_t(adjust(backlog.relative_size()) * delay_limit_us));
-        // "budget" has millisecond resolution and can potentially be long
-        // in the future so converting it to microseconds may overflow.
-        // So to compare buget and ret we need to convert both to the lower
-        // resolution.
-        if (std::chrono::duration_cast<storage_proxy::clock_type::duration>(ret) < budget) {
-            return ret;
-        } else {
-            // budget is small (< ret) so can be converted to microseconds
-            return std::chrono::duration_cast<std::chrono::microseconds>(budget);
-        }
+
+        // The maximum delay applied to the response is the amount of time left until the request timeouts.
+        // Delaying the response more than the timeout point doesn't make much sense - the request will fail
+        // at the timeout point anyway.
+        std::chrono::nanoseconds budget =
+            std::max(std::chrono::nanoseconds(0), _expire_timer.get_timeout() - storage_proxy::clock_type::now());
+
```

Comment:
Why use nanoseconds instead of duration_type?

I don't understand the problem the patch is fixing.

Reply:
The original code looked like this:
```c++
std::chrono::microseconds calculate_delay(db::view::update_backlog backlog) {
    constexpr auto delay_limit_us = 1000000;
    auto adjust = [] (float x) { return x * x * x; };
    auto budget = std::max(storage_proxy::clock_type::duration(0),
        _expire_timer.get_timeout() - storage_proxy::clock_type::now());
    std::chrono::microseconds ret(uint32_t(adjust(backlog.relative_size()) * delay_limit_us));
    // "budget" has millisecond resolution and can potentially be long
    // in the future so converting it to microseconds may overflow.
    // So to compare buget and ret we need to convert both to the lower
    // resolution.
    if (std::chrono::duration_cast<storage_proxy::clock_type::duration>(ret) < budget) {
        return ret;
    } else {
        // budget is small (< ret) so can be converted to microseconds
        return std::chrono::duration_cast<std::chrono::microseconds>(budget);
    }
}
```
`storage_proxy::clock_type::duration` is an alias for `std::chrono::nanoseconds`.

The original code assumed that `storage_proxy::clock_type::duration` is `std::chrono::millisconds`, not `nanoseconds`.
It calculates the amount of time to sleep for, the result is `std::chrono::microseconds`.
There's also a `budget` - maximum amount of time that we are allowed to sleep for. Budget is expressed as `storage_proxy::clock_type::duration`, which the original code assumed to be `milliseconds`, but is actually `nanoseconds`.

Directly comparing `milliseconds` and `microseconds` is risky - casting `int64::max()` of milliseconds to microseconds would lead to overflow, as the cast multiplies the amount of milliseconds by 1000.

To avoid overflow the original code did the comparison in a safe way:
```c++
(milliseconds)ret < budget
```

Casting the ret, which is in `microseconds` to budget's type, `milliseconds` is safe - this is a division by 1000.

But the problem is that the budget is actually in `nanoseconds`, not `milliseconds`.
Because of this the anti-overflow comparison becomes:
```c++
(nanoseconds)ret < budget
```
This is unsafe - ret is in `microseconds`, so casting it to `nanoseconds` means multiplying by 1000, which could overflow.

The commit fixes the issue by casting the `budget` from `nanoseconds` to `microseconds` (division by 1000, safe) and then doing the comparison.

I used the `std::chrono::nanoseconds` directly instead of the alias to signify that the value is expected to be in `nanoseconds`. If somebody changes `storage_proxy::clock_type::duration` to be something else this part of code won't compile and they will have to take a look at this part of the code and modify overflow protection to handle the new type.

Reply:
Please extract it into a separate pull request with an issue if you think it's fixing a bug. We also have to think about how this is tested.

Reply:
But I don't see how the original code can fail. It's an expiration time. Even if many years, it will fit in 64 bits.

Reply:
> Please extract it into a separate pull request with an issue if you think it's fixing a bug. We also have to think about how this is tested.

Opened https://github.com/scylladb/scylladb/pull/14637